ironmon_tracker/data/TrainerData.lua

```diff
@@ -75,17 +124,113 @@ function TrainerData.initialize()
-		local randomIcon = Utils.inlineIf(math.random(2) == 1, "a", "b") -- For now, random between female and male trainer
-		return {
-			name = "Unknown",
-			group = TrainerData.TrainerGroups.Other,
-			filename = "unknown-" .. randomIcon,
+	return TrainerData.Trainers[trainerId or false] or TrainerData.BlankTrainer
+end
+
+-- Determines if the trainer's display name should use its in-game name (i.e. Terry) or class name (i.e. Rival)
+function TrainerData.shouldUseClassName(trainerId)
+	if trainerId == nil then return false end
+	if GameSettings.game == 3 then
+		if 326 <= trainerId and trainerId <= 334 then -- Rivals
+			return true
+		elseif 426 <= trainerId and trainerId <= 440 then -- Rivals
+			return true
+		elseif 739 <= trainerId and trainerId <= 741 then -- Rivals
+			return true
+		end
+	end
+	return false
+end
+
+-- Determines if this trainer should be used; if it's a rival then it has to be the correct rival for the player
+function TrainerData.shouldUseTrainer(trainerId)
+	local trainerInternal = TrainerData.getTrainerInfo(trainerId) or {}
+	if trainerInternal == TrainerData.BlankTrainer then
+		return false
+	end
+	-- Always okay to use trainer if it's not a rival
+	if Tracker.Data.whichRival == nil or trainerInternal.whichRival == nil then
+		return true
+	end
+	-- Otherwise, make sure it's the correct rival
+	return Tracker.Data.whichRival == trainerInternal.whichRival
+end
+
+-- Returns a list of trainers to exclude from a parsed log, often dummy trainers or VS Seeker rematch
+function TrainerData.getExcludedTrainers()
+	-- Holds individual trainerIds or a table-pair of ids, which is a range of ids
+	local trainerIdRanges = {}
+	if GameSettings.game == 1 or GameSettings.game == 2 then -- Ruby/Sapphire/Emerald
+		trainerIdRanges = {
+			{40, 43}, {47, 50}, {54, 56}, {60, 63}, {67, 70}, {84, 87}, {101, 104}, {110, 113}, 117,
+			{120, 123}, {132, 135}, {139, 142}, {147, 150}, 173, {175, 178}, {184, 187}, {197, 200},
+			{207, 210}, {219, 222}, {228, 231}, {239, 242}, {250, 253}, {257, 260}, {276, 279}, {282, 285},
+			{288, 291}, {295, 298}, {303, 306}, {308, 311}, {314, 317}, {328, 331}, 341, {346, 349},
+			{354, 357}, {360, 363}, {365, 368}, {370, 373}, {379, 382}, {388, 391}, {393, 396}, {409, 412},
+			{421, 424}, {430, 433}, {437, 440}, 456, 462, {466, 468}, {477, 480}, 482, {485, 489},
+			{497, 500}, {515, 518}, {541, 544}, {548, 551}, {555, 558}, {562, 565}, {607, 610}, {622, 625},
+			{633, 634}, {636, 639}, {643, 646}, {657, 660}, {682, 685}, {688, 691}, {770, 801}, {805, 847},
+			{851, 855},
 		}
+	elseif GameSettings.game == 3 then -- FireRed/LeafGreen
+		trainerIdRanges = {
+			{1, 88}, 147, 200, 263, {492, 497}, {511, 515}, 530,
+		}
+	end
+
+	local trainerIds = {}
+	for _, item in pairs(trainerIdRanges) do
+		if type(item) == "number" then
+			trainerIds[item] = true
+		elseif type(item) == "table" then
+			for i = (item[1] or 9999), (item[2] or 9999), 1 do
```

Comment:
do we need the fallback defaults here, considering that the ranges are just defined right above? (so we know there *shouldn't* be any nil or missing items unless there was user error in whomever is editing the `trainerIdRanges`)

In this case, if we're factoring in possibilities for dev error on the array defined above, there could technically be an issue where if `trainerIdRanges` contains any value like `{ idX }` (i.e. is a table but with one value) then we wind up filling in values for `trainerIds` from `idX` to `9999` (including trainers that we wouldn't want to exclude) which may / may not be as obvious initially as having an error from nil / missing value

Reply:
Good catch, and yes that makes total sense. I'll remove the defaults and just add a slightly safer check to the entry condition.

Fixed in https://github.com/besteon/Ironmon-Tracker/pull/403/commits/2b90717aaba420a65feeedf24abcc4e88f826be0